metricbeat/module/mongodb/oplog/oplog.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/elastic/beats/metricbeat/module/mongodb"
+)
+
+const oplogCol = "oplog.rs"
+
+var debugf = logp.MakeDebug("mongodb.oplog")
+
+func init() {
+	mb.Registry.MustAddMetricSet("mongodb", "oplog", New,
+		mb.WithHostParser(mongodb.ParseURL),
+		mb.DefaultMetricSet())
+}
+
+// MetricSet type defines all fields of the MetricSet
+// As a minimum it must inherit the mb.BaseMetricSet fields, but can be extended with
+// additional entries. These variables can be used to persist data or configuration between
+// multiple fetch calls.
+type MetricSet struct {
+	*mongodb.MetricSet
+}
+
+func contains(s []string, x string) bool {
+	for _, n := range s {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// New creates a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The mongodb oplog metricset is experimental.")
+
+	ms, err := mongodb.NewMetricSet(base)
+	if err != nil {
+		return nil, err
+	}
+	return &MetricSet{ms}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch() (common.MapStr, error) {
+	// instantiate direct connections to each of the configured Mongo hosts
+	mongoSession, err := mongodb.NewDirectSession(m.DialInfo)
+	if err != nil {
+		return nil, err
+	}
+	defer mongoSession.Close()
+
+	// get oplog.rs collection
+	db := mongoSession.DB("local")
+	if collections, err := db.CollectionNames(); err != nil || !contains(collections, oplogCol) {
+		if err == nil {
+			err = errors.New("Collection oplog.rs was not found")
+		}
+
+		logp.Err(err.Error())
+		return nil, err
+	}
+	collection := db.C(oplogCol)
+
+	//  oplog size
+	var oplogStatus map[string]interface{}
+	if err := db.Run(bson.D{{Name: "collStats", Value: oplogCol}}, &oplogStatus); err != nil {
+		return nil, err
+	}
+
+	allocated := oplogStatus["maxSize"].(int64)
+	used := int64(oplogStatus["size"].(float64))
+
+	// get first and last items in the oplog
+	oplogIter := collection.Find(nil).Sort("$natural").Iter()
+	oplogReverseIter := collection.Find(nil).Sort("-$natural").Iter()
+	var first, last interface{}
+	if !oplogIter.Next(&first) || !oplogReverseIter.Next(&last) {
+		err := errors.New("Objects not found in local.oplog.rs -- Is this a new and empty db instance?")
+		logp.Err(err.Error())
+		return nil, err
+	}
+
+	firstTsValue, firstOk := first.(bson.M)["ts"].(bson.MongoTimestamp)
+	lastTsValue, lastOk := last.(bson.M)["ts"].(bson.MongoTimestamp)
+	if !firstOk || !lastOk {
+		err := errors.New("Unexpected timestamp value found in first/last oplog item")
+		return nil, err
+	}
+	firstTs := int64(firstTsValue)
+	lastTs := int64(lastTsValue)
```

Comment:
Logic to get and validate first and last timestamps is mostly the same, you could create a function to get them, it could be a function with parameters for the `collection` and the sorting order (`$natural`/`-$natural`), so at the end you can have here something like:
```
first, err := getTimestamp(collection, "$natural")
if err != nil {
    return nil, errors.Wrap(err, "failed to obtain timestamp of first operation")
}
last, err := getTimestamp(collection, "-$natural")
if err != nil {
    return nil, errors.Wrap(err, "failed to obtain timestamp of last operation")
}
diff := last - first
```

Reply:
I tried to define: `func getTimestamp(collection *mgo.Collection, sort string) (int64, error)` but there was an issue with referencing the `*Collection` type:
```
cannot use collection (type *"github.com/elastic/beats/vendor/gopkg.in/mgo.v2".Collection) as
 type *"gopkg.in/mgo.v2".Collection in argument to getTimestamp
```
I don't know how to reference a type from the package mgo.v2 which is a vendor package of the beats platform.

Reply:
Umm, it looks fine, I downloaded your branch and it compiles, this module is also used in `mongodb/mongodb.go` :thinking: